library/core/src/panic/assert_info.rs

```diff
@@ -0,0 +1,52 @@
+use crate::fmt::Debug;
+
+/// Information about a failed assertion.
+#[derive(Debug)]
+pub struct AssertInfo<'a> {
+    /// The assertion that failed.
+    pub assertion: Assertion<'a>,
+    /// Optional additional message to include in the failure report.
+    pub message: Option<&'a crate::fmt::Arguments<'a>>,
+}
+
+#[derive(Debug)]
+pub enum Assertion<'a> {
+    /// The assertion is a boolean assertion.
+    Bool(BoolAssertion),
```

Comment:
Maybe clarify that this is `assert!(..)`, and/or that the boolean was asserted to be `true`. (And same for the `BoolAssertion` type below.)

Reply:
When done, `assert!()` should also be able fail with `BinaryAssertion`. I'll use the comments to explain it is meant for boolean expressions that can't be parsed/split up in a more descriptive manner.